presto-router/src/main/java/com/facebook/presto/router/cluster/ClusterManager.java

```diff
@@ -21,88 +22,360 @@
-        this.groupSelectors = ImmutableList.copyOf(routerSpec.getSelectors());
-        this.schedulerType = routerSpec.getSchedulerType();
-        this.scheduler = new SchedulerFactory(routerSpec.getSchedulerType()).create();
+        Map<URI, URI> newDiscoveryURIs = new HashMap<>();
+        initializeMembersDiscoveryURI(newDiscoveryURIs, newGroups);
 
-        this.initializeServerWeights();
+        updatedAllClusters.forEach(uri -> {
+            remoteClusterInfos.computeIfAbsent(uri, value -> remoteInfoFactory.createRemoteClusterInfo(newDiscoveryURIs.get(value)));
+            remoteQueryInfos.computeIfAbsent(uri, value -> remoteInfoFactory.createRemoteQueryInfo(newDiscoveryURIs.get(value)));
+            log.info("Attached cluster %s to the router. Queries will be routed to cluster after successful health check", uri.getHost());
+        });
+
+        for (URI uri : remoteClusterInfos.keySet()) {
+            if (!updatedAllClusters.contains(uri)) {
+                remoteClusterInfos.remove(uri);
+                remoteQueryInfos.remove(uri);
+                log.info("Removed cluster %s from the router", uri.getHost());
+            }
+        }
+        currentConfig.set(new ClusterManagerConfig(newGroups, newGroupSelectors, newScheduler, newSchedulerType));
+    }
+
+    @PostConstruct
+    public void startConfigReloadTaskFileWatcher()
+    {
+        CompletableFuture.supplyAsync(() -> {
+            try (WatchService watchService = FileSystems.getDefault().newWatchService()) {
+                File routerConfigFile = new File(routerConfig.getConfigFile());
+                log.info("Router config watch service monitoring %s", routerConfig.getConfigFile());
+                Path parentDir = routerConfigFile.toPath().getParent();
+                parentDir.register(
+                        watchService,
+                        new WatchEvent.Kind[] {
+                                StandardWatchEventKinds.ENTRY_MODIFY,
+                                StandardWatchEventKinds.ENTRY_CREATE,
+                                StandardWatchEventKinds.ENTRY_DELETE,
+                                StandardWatchEventKinds.OVERFLOW},
+                        SensitivityWatchEventModifier.HIGH);
+                isWatchServiceStarted.set(true);
+                log.info("Successfully registered watch service for %s", parentDir);
+
+                while (true) {
+                    WatchKey key = watchService.take();
+                    log.info("Changes to router config directory detected: %s", routerConfigFile);
+                    for (WatchEvent<?> event : key.pollEvents()) {
+                        log.debug("Event detected: %s, path: %s", event.kind().name(), event.context());
+                        Path changed = (Path) event.context();
+                        if (changed.endsWith(routerConfigFile.getName())) {
+                            onConfigChangeDetection();
```

Comment:
If this ever throws an exception, the loop will exit. This could easily happen with a bad router config. Should we catch exceptions here?